### PR TITLE
Fixes issue 816 (space in DS names) and add tenant name validation during tenant create

### PR DESCRIPTION
**Pull request by lipingxue**
This PR includes:
1. Fix issue 816(space in DS names)
2. validate tenant name during tenant create, tenant name are only allowed to be [a-zA-Z0-9_][a-zA-Z0-9_.-]*

### CODE REVIEW

File: esx_service/vmdk_ops.py [hunk 1]
```diff
@@ -854,11 +854,14 @@ def findDeviceByPath(vmdk_path, vm):
         # to match with the given volume name.
         # Filename format is as follows:
         #   "[<datastore name>] <parent-directory>/tenant/<vmdk-descriptor-name>"
-        logging.debug("d.backing.fileName %s", d.backing.fileName)
-        backing_disk = d.backing.fileName.split(" ")[1]
+        # Assume there is one space between these two parts "[<datastore name>]" and
+        # <parent-directory>/tenant/<vmdk-descriptor-name>
+        logging.debug("findDeviceByPath: d.backing.fileName %s", d.backing.fileName)
+        backing_disk = d.backing.fileName.rsplit(" ", 1)[1]
```
Comment (msterin) [inline esx_service/vmdk_ops.py]: Can  `<parent-directory>/tenant/<vmdk-descriptor-name>` hasve a space ? The current code is guaranteed to fail. Maybe it's better to look for ']' ? Something like
```python
    ds, file  = d.backing.fileName.rsplit("]", 1)[1]
    datastore = ds[1:]
    backing_disk = file[1:]
```
That raises a couple of additional things to check:
(1) can we create a tenant with a space 
(2) can we create a volume with a space in the name.

Comment (shaominchen) [inline esx_service/vmdk_ops.py]: IMO, we shouldn't allow spaces in tenant & volume names. It should be OK to have spaces in tenant description though.

File: esx_service/vmdk_ops.py [hunk 2]
```diff
@@ -854,11 +854,14 @@ def findDeviceByPath(vmdk_path, vm):
         # to match with the given volume name.
         # Filename format is as follows:
         #   "[<datastore name>] <parent-directory>/tenant/<vmdk-descriptor-name>"
-        logging.debug("d.backing.fileName %s", d.backing.fileName)
-        backing_disk = d.backing.fileName.split(" ")[1]
+        # Assume there is one space between these two parts "[<datastore name>]" and
```
Comment (shaominchen) [inline esx_service/vmdk_ops.py]: This comment is not very clear to me. I'd suggest to give an example in the comment so that it's easier to understand the logic of ....rsplit(" ", 1)[0]

File: esx_service/vmdk_ops.py [hunk 3]
```diff
@@ -854,11 +854,14 @@ def findDeviceByPath(vmdk_path, vm):
...
+        # <parent-directory>/tenant/<vmdk-descriptor-name>
+        logging.debug("findDeviceByPath: d.backing.fileName %s", d.backing.fileName)
+        backing_disk = d.backing.fileName.rsplit(" ", 1)[1]
 
         # datastore='[datastore name]'
```
Comment (shaominchen) [inline esx_service/vmdk_ops.py]: Same as above: suggest to add an example in the comment to explain the logic below

File: esx_service/vmdk_ops.py (hunk 1 above)
Comment (lipingxue) [inline esx_service/vmdk_ops.py]: I don't know space in "volume names" is allowed in vsphere or not? "datastore" names can have space.

Comment (msterin) [inline esx_service/vmdk_ops.py]: That was the point - please check it :-). Directory names (i.e. tenant) and volume names (i.e. files) will accept spaces unless we block them on creation. So please check if we do block it on creation.  I agree with @shaominchen that it's better to block spaces in tenant names and volume names. on creation.
With that said, i still think the code I suggested is slightly more resilient than looking for spaces.

Comment (lipingxue) [inline esx_service/vmdk_ops.py]: Yes. I agree with you.  I will modify the code as you suggested. Currently, I think we don't check for tenant name and volume name during creation. I will add the check logic.

Comment (lipingxue) [inline esx_service/vmdk_ops.py]: I did some test.
Vsphere allows volume name including space since I can create a VM with name including space.

But if I create a volume using local driver, it does not allow volume name to include space, and I got the following error.

**Comment by govint** [comment]
@lipingxue can you also include a test for the scenario in #892, if spaces are disallowed then that issue shouldn't be able to occur.

What about is a VSphere has already created a datastore with spaces in it. Then will that datastore be prevented from being used in any tenant. Ideally, user should be disallowed from adding that datastore.

File: esx_service/utils/vmdk_utils.py [hunk 4]
```diff
@@ -232,7 +232,7 @@ def list_vmdks(path, volname="", show_snapshots=False):
 def vmdk_is_a_descriptor(path, file_name):
     """
     Is the file a vmdk descriptor file?  We assume any file that ends in .vmdk,
-    does not have -delta or -flat or he likes at the end of filename, 
+    does not have -delta or -flat or he likes at the end of filename,
```
Comment (govint) [inline esx_service/utils/vmdk_utils.py]: seems a typo, sentence must be corrected - "or he likes at the end of filename"

File: esx_service/vmdk_ops.py (hunk 1 above)
Comment (lipingxue) [inline esx_service/vmdk_ops.py]: Added code to check tenant name and volume name during creation.

File: esx_service/vmdk_ops.py (hunk 2 above)
Comment (lipingxue) [inline esx_service/vmdk_ops.py]: Done.

File: esx_service/vmdk_ops.py (hunk 3 above)
Comment (lipingxue) [inline esx_service/vmdk_ops.py]: Done.

File: esx_service/utils/vmdk_utils.py (hunk 4 above)
Comment (lipingxue) [inline esx_service/utils/vmdk_utils.py]: @govint We only disallow space in tenant name and volume name during creation. So with this fix, issue #892 can still happen. We cannot control the datastore name. In VSphere, spaces in datastore name are allowed.

Comment (lipingxue) [inline esx_service/utils/vmdk_utils.py]: Typo in comment is fixed too.

**Comment by lipingxue** [comment]
Add code to validate tenant_name and volume name during tenant create and volume create.
1. create a tenant with invalid name
```
[root@localhost:~] /usr/lib/vmware/vmdkops/bin/vmdkops_admin.py tenant create --name=.abc*d
Tenant name .abc*d is invalid, only [a-zA-Z0-9_][a-zA-Z0-9_.-]* is allowed

```

2. create a volume with invalid name
```
root@photon-eZ5ILvREQ [ ~ ]# docker volume create --driver=vmdk --name="abc*   edf"
Error response from daemon: create abc*   edf: VolumeDriver.Create: Volume name abc*   edf is invalid, only [a-zA-Z0-9][a-zA-Z0-9_.-]* is allowed
```

**Comment by lipingxue** [comment]
@msterin @shaominchen @govint I have addressed your comments, change is repushed.  Please review it again.

File: esx_service/utils/vmdk_utils.py [hunk 5]
```diff
@@ -100,8 +100,11 @@ def get_datastores():
     'url-name' is the last element of datastore URL (e.g. 'vsan:572904f8c031435f-3513e0db551fcc82')
     'dockvol-path; is a full path to 'dockvols' folder on datastore
     """
-    if datastores == None:
-        init_datastoreCache()
+    # if datastores == None:
+    #     init_datastoreCache()
+    # we should not use cached value, user may change datastore name on the fly
```
Comment (msterin) [inline esx_service/utils/vmdk_utils.py]: the problem is that init_datastoreCache is a an  expensive operation and it is called way too many times, often many times in the same call chain. The correct way to fix it would have been passing datastore list, but it was a very invasive change so the cache was introduced.  Let's discuss how to test and what are the other ways to fix the issue, rather than refetch all DS list every time

File: esx_service/utils/vmdk_utils.py [hunk 6]
```diff
@@ -100,8 +100,11 @@ def get_datastores():
     'url-name' is the last element of datastore URL (e.g. 'vsan:572904f8c031435f-3513e0db551fcc82')
     'dockvol-path; is a full path to 'dockvols' folder on datastore
     """
-    if datastores == None:
-        init_datastoreCache()
+    # if datastores == None:
+    #     init_datastoreCache()
```
Comment (msterin) [inline esx_service/utils/vmdk_utils.py]: Why this change is introduced ? It seems to irrelevant to #816

File: esx_service/utils/auth_api.py [hunk 7]
```diff
@@ -24,6 +24,10 @@
 import logging
 from error_code import ErrorCode
 from error_code import ErrorInfo
+import re
+
+# regex for valid tenant name
+VALID_TENANT_NAME_REG = "[a-zA-Z0-9_][a-zA-Z0-9_.-]*"
```
Comment (msterin) [inline esx_service/utils/auth_api.py]: Need to add documentation for this, a one liner

File: esx_service/utils/vmdk_utils.py [hunk 8]
```diff
@@ -100,8 +100,11 @@ def get_datastores():
...
+    # we should not use cached value, user may change datastore name on the fly
+    # need to repull the info
+    init_datastoreCache()
```
Comment (shaominchen) [inline esx_service/utils/vmdk_utils.py]: With this change, we are losing the purpose of introducing datastore cache. Have we filed an issue to fix the datastore cache implementation?

File: esx_service/utils/auth_api.py [hunk 9]
```diff
@@ -239,9 +243,22 @@ def get_default_datastore(name):
         error_info = error_code.generate_error_info(ErrorCode.INTERNAL_ERROR, error_msg)
     return error_info, default_datastore
 
+def is_valid_tenant_name(name):
+    """ Check the given name is a valid tenant name """
+    pattern = "^" + VALID_TENANT_NAME_REG + "$"
```
Comment (msterin) [inline esx_service/utils/auth_api.py]: nit: why not pre-compile it once and keep the compiled regexp in a global ? It's not a bit deal since the operation is rare, but generally it's a waste.

File: esx_service/utils/auth_api.py [hunk 10]
```diff
@@ -239,9 +243,22 @@ def get_default_datastore(name):
         error_info = error_code.generate_error_info(ErrorCode.INTERNAL_ERROR, error_msg)
     return error_info, default_datastore
 
+def is_valid_tenant_name(name):
+    """ Check the given name is a valid tenant name """
```
Comment (shaominchen) [inline esx_service/utils/auth_api.py]: Nit: "Check..." => "Check if..."

File: esx_service/utils/auth_api.py [hunk 11]
```diff
@@ -239,9 +243,22 @@ def get_default_datastore(name):
...
+    pattern = "^" + VALID_TENANT_NAME_REG + "$"
+    valid_name_reg = re.compile(pattern)
+    if valid_name_reg.match(name):
```
Comment (shaominchen) [inline esx_service/utils/auth_api.py]: Nit: we can simplify it a little bit by removing the if-else block:

   return valid_name_reg.match(name)

File: esx_service/vmdk_ops.py [hunk 12]
```diff
@@ -153,12 +157,24 @@ def RunCommand(cmd):
...
+        return True
+    else:
+        return False
 
 # returns error, or None for OK
 # opts is  dictionary of {option: value}.
 # for now we care about size and (maybe) policy
 def createVMDK(vmdk_path, vm_name, vol_name, opts={}, vm_uuid=None, tenant_uuid=None, datastore=None):
     logging.info("*** createVMDK: %s opts = %s", vmdk_path, opts)
+    if not is_valid_volume_name(vol_name):
```
Comment (msterin) [inline esx_service/vmdk_ops.py]: I think we already have  code to parse volume names - please check if can re-use

File: esx_service/vmdk_ops.py [hunk 13]
```diff
@@ -153,12 +157,24 @@ def RunCommand(cmd):
 
     return (s, o)
 
+def is_valid_volume_name(name):
+    """ Check the given name is a valid volume name """
```
Comment (shaominchen) [inline esx_service/vmdk_ops.py]: Nit: "Check..." => "Check if..."

File: esx_service/vmdk_ops.py [hunk 14]
```diff
@@ -153,12 +157,24 @@ def RunCommand(cmd):
...
+    pattern = "^" + VALID_VOLUME_NAME_REG + "$"
+    valid_name_reg = re.compile(pattern)
+    if valid_name_reg.match(name):
```
Comment (shaominchen) [inline esx_service/vmdk_ops.py]: Nit: Same as above - we can simplify it a little bit:

   return valid_name_reg.match(name)

File: esx_service/utils/vmdk_utils.py (hunk 6 above)
Comment (lipingxue) [inline esx_service/utils/vmdk_utils.py]: This change is needed. Think about this case:
step 1: assume we have datastore named "datastore1", which is the default_datastore, we create a volume "vol1"
step 2: rename "datastore1" to "datastore 1" (including space in the name)
step3: create another vol "vol2", which will fail.  

It will hit the error same as #818 . The root cause is "vmdk_utils.get_datastores" returns cached value which is outdated.

File: esx_service/utils/vmdk_utils.py (hunk 8 above)
Comment (lipingxue) [inline esx_service/utils/vmdk_utils.py]: I discussed with Mark offline, we decided to remove this cache related change from this PR.

File: esx_service/utils/vmdk_utils.py (hunk 5 above)
Comment (lipingxue) [inline esx_service/utils/vmdk_utils.py]: As we discussed offline, will remove this change from this PR.

File: esx_service/vmdk_ops.py (hunk 12 above)
Comment (lipingxue) [inline esx_service/vmdk_ops.py]: We have a util function parse_vol_name, which check whether the vol_name is match r"^.*-[0-9]{6}$" . So there is no limit on which character will be included in the vol_name. But docker local driver limit vol_name can only be "[a-zA-Z0-9][a-zA-Z0-9_.-]" . Do we want have the same limit as docker local driver does?

Comment (lipingxue) [inline esx_service/vmdk_ops.py]: @msterin

File: esx_service/vmdk_ops.py (hunk 13 above)
Comment (lipingxue) [inline esx_service/vmdk_ops.py]: Done. Remove this function and introduce a global variable "valid_volume_name_reg", use valid_volume_name_reg.match() to check directly.

File: esx_service/vmdk_ops.py (hunk 14 above)
Comment (lipingxue) [inline esx_service/vmdk_ops.py]: Done.

File: esx_service/vmdk_ops.py [hunk 15]
```diff
@@ -129,6 +130,9 @@
 # For managing resource locks.
 lockManager = threadutils.LockManager()
 
+# regex for valid volume name
+VALID_VOLUME_NAME_REG = "[a-zA-Z0-9][a-zA-Z0-9_.-]*"
```
Comment (pdhamdhere) [inline esx_service/vmdk_ops.py]: Why do we need this one? We already validate volume name in `parse_vol_name`.

File: esx_service/utils/vmdk_utils.py (hunk 8 above)
Comment (pdhamdhere) [inline esx_service/utils/vmdk_utils.py]: Are you going to remove `init_datastoreCache()`

File: esx_service/vmdk_ops.py (hunk 15 above)
Comment (lipingxue) [inline esx_service/vmdk_ops.py]: docker local driver has this limitation on volume names during volume create. I think we should apply the same check.  We can discuss if you think differently. @msterin

**Comment by govint** [comment]
When trying out with a vol@<data store name with spaces> figured the below:

1. Run command to mkdir - failed because of spaces, quoting the path name
passed to mkdir worked.

2. find_devicebypath() in vmdk_ops.py has some code to get the datastore
name from a vmdk path and thats splitting the path incorrectly when the
datastore name has spaces in it.

Basically, fix an issue and then it breaks elsewhere. Keep fixing the
issues till the entire use case (create/get/remove/attach/detach) works
fine.

With (1) above fixed create is ok, attach/detach fail in (2) above.

File: esx_service/vmdk_ops.py (hunk 15 above)
Comment (msterin) [inline esx_service/vmdk_ops.py]: We do not care what local diver does, so unless there is a good reason for  the change allowed volnames, please remove this code. 

And, as Prashant has mentioned, the parsing is happening in `parse_vol_name` so any changes should be done there.

File: esx_service/vmdk_ops.py (hunk 12 above)
Comment (msterin) [inline esx_service/vmdk_ops.py]: parse_vol_name does more that that - it also separated DS name and handles them differnetly.
And no, I see no reason to mimic local local driver, we have deviated from it already by adding `@` to full volume name.

File: esx_service/utils/vmdk_utils.py (hunk 8 above)
Comment (lipingxue) [inline esx_service/utils/vmdk_utils.py]: No, we will keep init_datastoreCache().

**Comment by lipingxue** [comment]
@msterin I have addressed your comments:
1. remove "whitespace cleanup" from this PR
2. remove the volume name checking during volume creation

Please review it.

**Comment by lipingxue** [comment]
Since Sam is on PTO, @msterin @govint @pdhamdhere please review it.

File: esx_service/utils/auth_api.py [hunk 16]
```diff
@@ -241,7 +247,12 @@ def get_default_datastore(name):
 
 def _tenant_create(name, description="", vm_list=None, privileges=None):
     """ API to create a tenant """
+    global valid_tenant_name_reg
     logging.debug("_tenant_create: name=%s description=%s vm_list=%s privileges=%s", name, description, vm_list, privileges)
+    if not valid_tenant_name_reg.match(name):
+        error_info = error_code.generate_error_info(ErrorCode.TENANT_NAME_INVALID, name, VALID_TENANT_NAME_REGEXP)
+        return error_info, None
+
```
Comment (govint) [inline esx_service/utils/auth_api.py]: Can this be a function like is_tenant_name_valid() vs. repeated in multiple locations.

File: esx_service/vmdk_ops_test.py [hunk 17]
```diff
@@ -715,20 +716,27 @@ def create_default_tenant_and_privileges(self):
...
+            if error_info:
+                logging.warning(error_info.msg)
 
         # create DEFAULT privilege if needed
-        error_info, privileges = auth.get_default_privileges()
+        error_msg, privileges = auth.get_default_privileges()
+        if error_msg:
+            logging.warning(error_msg)
+            return
+
```
Comment (govint) [inline esx_service/vmdk_ops_test.py]: This action and error handling must be done before tenant create, its leaving a new tenant created and left in an unknown state (line 726 above).

File: docs/user-guide/admin-cli.md [hunk 18]
```diff
@@ -39,17 +39,17 @@ optional arguments:
 ### Create
 A tenant named "_DEFAULT" will be created automatically post install.
 
-Creates a new named tenant and optionally assigns VMs.
+Creates a new named tenant and optionally assigns VMs. Valid tenant name is only allowed to be "[a-zA-Z0-9_][a-zA-Z0-9_.-]*"
```
Comment (shuklanirdesh82) [inline docs/user-guide/admin-cli.md]: @lipingxue  a quick question: PR description says ```Fixes issue 816 (space in DS names).``` can you please help to understand why some of the fixes are related with ```TENANT_NAME_INVALID``` or ```VOLUME_NAME_INVALID```?

File: esx_service/vmdk_ops_test.py [hunk 19]
```diff
@@ -715,20 +716,27 @@ def create_default_tenant_and_privileges(self):
         """ Create default tenant and privilege if not exist"""
 
         # create DEFAULT tenant if needed
-        error_info, tenant_uuid, tenant_name = auth.get_default_tenant()
+        error_msg, tenant_uuid, tenant_name = auth.get_default_tenant()
+        if error_msg:
+            logging.warning(error_msg)
```
Comment (shuklanirdesh82) [inline esx_service/vmdk_ops_test.py]: don't we need to fail the test here ( not able to create default tenant)?

File: esx_service/vmdk_ops.py [hunk 20]
```diff
@@ -855,11 +856,10 @@ def findDeviceByPath(vmdk_path, vm):
         # Filename format is as follows:
         #   "[<datastore name>] <parent-directory>/tenant/<vmdk-descriptor-name>"
         logging.debug("d.backing.fileName %s", d.backing.fileName)
-        backing_disk = d.backing.fileName.split(" ")[1]
-
-        # datastore='[datastore name]'
-        datastore = d.backing.fileName.split(" ")[0]
-        datastore = datastore[1:-1]
+        ds, disk_path = d.backing.fileName.rsplit("]", 1)
```
Comment (shuklanirdesh82) [inline esx_service/vmdk_ops.py]: can you please confirm about having a test for this code snip?

File: esx_service/utils/auth_api.py [hunk 21]
```diff
@@ -24,6 +24,12 @@
 import logging
 from error_code import ErrorCode
 from error_code import ErrorInfo
+import re
+
+# regex for valid tenant name
+VALID_TENANT_NAME_REGEXP = "[a-zA-Z0-9_][a-zA-Z0-9_.-]*"
```
Comment (shuklanirdesh82) [inline esx_service/utils/auth_api.py]: 1. Is it something we are introducing new?
2. What happens if customer is having tenant name that violates this rule? I guess new tenant creates enforces the restriction but would like to make sure here.

File: esx_service/utils/auth_api.py [hunk 22]
```diff
@@ -241,7 +247,12 @@ def get_default_datastore(name):
 
 def _tenant_create(name, description="", vm_list=None, privileges=None):
     """ API to create a tenant """
+    global valid_tenant_name_reg
     logging.debug("_tenant_create: name=%s description=%s vm_list=%s privileges=%s", name, description, vm_list, privileges)
+    if not valid_tenant_name_reg.match(name):
```
Comment (shuklanirdesh82) [inline esx_service/utils/auth_api.py]: 1. need to add testing done for this code snip
2. can you confirm about having tests exercising this new change?

File: esx_service/utils/auth_api.py (hunk 16 above)
Comment (lipingxue) [inline esx_service/utils/auth_api.py]: Done.

File: esx_service/vmdk_ops_test.py (hunk 17 above)
Comment (lipingxue) [inline esx_service/vmdk_ops_test.py]: Addressed your comments. Now the logic is
1. check default_tenant exists or not
2. if default_tenant exists, do nothing, continue
3. if default_tenant does not exist, try to create, assert the create should not return error, since the test need to start with default_tenant exist

Similar logic also apply to default_privilege

File: docs/user-guide/admin-cli.md (hunk 18 above)
Comment (lipingxue) [inline docs/user-guide/admin-cli.md]: This tenant name validation fix is not related to issue 816 itself.  Sam pointed out earlier that in his comment to suggest that we should give some limitation to the tenant name during creation. For example, no space allowed in the tenant name, which will avoid some potential issues that may occur like issue 816 if user specify tenant name with spaces and other weird characters.

File: esx_service/vmdk_ops_test.py (hunk 19 above)
Comment (lipingxue) [inline esx_service/vmdk_ops_test.py]: Yes. I addressed this. See above comments I replied to Govindan's comments.

File: esx_service/utils/auth_api.py (hunk 21 above)
Comment (lipingxue) [inline esx_service/utils/auth_api.py]: Yes, the tenant create with invalid name should fail.

File: docs/user-guide/admin-cli.md (hunk 18 above)
Comment (shuklanirdesh82) [inline docs/user-guide/admin-cli.md]: It would be helpful if you can add some useful information in the PR title as well as in the first comment. 

Such information will be useful for future reference while searching from title.  Current title proposing the ```Fixes issue 816 (space in DS names).```

**Comment by lipingxue** [comment]
root@photon-eZ5ILvREQ [ ~ ]# docker volume create --driver=vmdk --name=vol2
vol2
root@photon-eZ5ILvREQ [ ~ ]# docker volume ls                              
DRIVER              VOLUME NAME
vmdk                vol1@datastore    1
vmdk                vol2@datastore    1

**Comment by lipingxue** [comment]
Test to make sure vol can be created on a datastore which name includes spaces.
Configuration: 
only _DEFAULT tenant is configured, VM "photon5" is created on "datastore1"
Step1: From VM"photon5", create a volume "vol1"
```
root@photon-eZ5ILvREQ [ ~ ]# docker volume create --driver=vmdk --name=vol1
vol1
root@photon-eZ5ILvREQ [ ~ ]# docker volume ls
DRIVER              VOLUME NAME
vmdk                vol1@datastore1
```

Step2:
rename datastore "datastore1" to "datastore    1" (to insert spaces)
restart /etc/init.d/vmdk-opsd service. This is to avoid hitting the issue caused by stale datastore info returned by datastore cache(issue 818, which I am working on).  

Step3: create another volume "vol2"
```
root@photon-eZ5ILvREQ [ ~ ]# docker volume ls
DRIVER              VOLUME NAME
vmdk                vol1@datastore    1
root@photon-eZ5ILvREQ [ ~ ]# docker volume create --driver=vmdk --name=vol2
vol2
root@photon-eZ5ILvREQ [ ~ ]# docker volume ls                              
DRIVER              VOLUME NAME
vmdk                vol1@datastore    1
vmdk                vol2@datastore    1

```

**Comment by lipingxue** [comment]
Test to make sure create tenant with invalid name will fail

create a tenant with invalid name like "tenant     abc" failed  as expect.
```
[root@localhost:/vmfs/volumes/57473702-8b550986-70c0-000c290799f4/dockvols] /usr/lib/vmware/vmdkops/bin/vmdkops_admin.py tenant ls
Uuid                                  Name      Description               Default_datastore  VM_list  
------------------------------------  --------  ------------------------  -----------------  -------  
11111111-1111-1111-1111-111111111111  _DEFAULT  This is a default tenant                              
[root@localhost:/vmfs/volumes/57473702-8b550986-70c0-000c290799f4/dockvols] /usr/lib/vmware/vmdkops/bin/vmdkops_admin.py tenant create --name="tenant     abc"
Tenant name tenant     abc is invalid, only [a-zA-Z0-9_][a-zA-Z0-9_.-]* is allowed

```

File: docs/user-guide/admin-cli.md (hunk 18 above)
Comment (lipingxue) [inline docs/user-guide/admin-cli.md]: OK. Change the title.

File: esx_service/vmdk_ops.py (hunk 20 above)
Comment (lipingxue) [inline esx_service/vmdk_ops.py]: Yes. I added test for this.

**Comment by govint** [comment]
@lipingxue can you also add a test that creates the volume on a datastore with spaces in it like "docker volume create --driver=vmdk --name=vol1@data\ \ \ store"

**Comment by lipingxue** [comment]
@govint Added the test you request
Configuration: only _DEFAULT tenant is configured, VM "photon5" is created on datastore1
Step1: create two volumes "vol1" and "vol2"
```
root@photon-eZ5ILvREQ [ ~ ]# docker volume create --driver=vmdk --name=vol1
vol1
root@photon-eZ5ILvREQ [ ~ ]# docker volume create --driver=vmdk --name=vol2
vol2
root@photon-eZ5ILvREQ [ ~ ]# docker volume ls
DRIVER              VOLUME NAME
vmdk                vol1@datastore1
vmdk                vol2@datastore1

```
Step2:
rename datastore "datastore1" to "datastore   1"
restart /etc/init.d/vmdk-opsd service. This is to avoid hitting the issue caused by stale datastore info returned by datastore cache(issue 818, which I am working on).

Step3: create a volume "vol3@datastore   1"
```
root@photon-eZ5ILvREQ [ ~ ]# docker volume create --driver=vmdk --name="vol3@datastore   1"
vol3@datastore   1

```

Step4: create a volume "vol4@datastore\ \ \ 1" with the following command
```
root@photon-eZ5ILvREQ [ ~ ]# docker volume create --driver=vmdk --name=vol4@datastore\ \ \ 1
vol4@datastore   1
root@photon-eZ5ILvREQ [ ~ ]# docker volume ls
DRIVER              VOLUME NAME
vmdk                vol1@datastore   1
vmdk                vol2@datastore   1
vmdk                vol3@datastore   1
vmdk                vol4@datastore   1

```

**Comment by govint** [comment]
@lipingxue thanks and looks good.

File: esx_service/vmdk_ops.py [hunk 23]
```diff
@@ -87,6 +87,7 @@
 import error_code
 import auth_api
 from error_code import ErrorCode
+import re
```
Comment (shuklanirdesh82) [inline esx_service/vmdk_ops.py]: trivial: I guess this needs to be removed as well.

Comment (shuklanirdesh82) [inline esx_service/vmdk_ops.py]: never mind .. not sure why diff is saying it is added as part of this change. Please ignore above comment. Sorry for the confusion.

Thanks!

Comment (lipingxue) [inline esx_service/vmdk_ops.py]: Done. Thanks for review.
